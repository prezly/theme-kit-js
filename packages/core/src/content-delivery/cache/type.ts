export type UnixTimestampInSeconds = number;

type Awaitable<T> = T | Promise<T>;

export interface Cache {
    get<T>(key: string, latestVersion: UnixTimestampInSeconds): Awaitable<T | undefined>;
    set<T>(key: string, value: T, version: UnixTimestampInSeconds): Awaitable<void>;
}
