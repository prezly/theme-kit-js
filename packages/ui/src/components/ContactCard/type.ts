import type { NewsroomContact } from '@prezly/sdk';

export type ContactInfo = Pick<
    NewsroomContact,
    | 'avatar_image'
    | 'name'
    | 'description'
    | 'company'
    | 'email'
    | 'phone'
    | 'mobile'
    | 'website'
    | 'facebook'
    | 'twitter'
>;
