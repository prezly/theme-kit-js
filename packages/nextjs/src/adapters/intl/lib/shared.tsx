/* eslint-disable @typescript-eslint/no-use-before-define,react/jsx-props-no-spreading */
import type { ReactElement, TimeHTMLAttributes } from 'react';
import { Fragment } from 'react';

import { DEFAULT_DATE_FORMAT, DEFAULT_TIME_FORMAT } from './constants';
import type {
    DateFormat,
    IntlDictionary,
    IntlMessageDescriptor,
    IntlMessageValues,
    Iso8601Date,
    TimeFormat,
    UnixTimestampInSeconds,
} from './types';

export function formatMessageString(
    dictionary: IntlDictionary,
    { id, defaultMessage }: IntlMessageDescriptor,
    values: IntlMessageValues<string> = {},
) {
    const format = dictionary[id];

    if (!format) {
        return defaultMessage ?? `[${id}]` ?? '';
    }

    return format
        .map(({ type, value }) => {
            if (type === 1) {
                return values[value] ?? `[${value}]`;
            }

            return value;
        })
        .join('');
}

export function formatMessageFragment(
    dictionary: IntlDictionary,
    { id, defaultMessage }: IntlMessageDescriptor,
    values: IntlMessageValues<string | ReactElement> = {},
): ReactElement {
    const format = dictionary[id];

    if (!format) {
        return <>{defaultMessage ?? `[${id}]` ?? ''}</>;
    }

    return (
        <>
            {format.map(({ type, value }, index) => {
                if (type === 1) {
                    return <Fragment key={index}>{values[value] ?? `[${value}]`}</Fragment>;
                }

                return <Fragment key={index}>{value}</Fragment>;
            })}
        </>
    );
}

type Month = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;

const MONTH_NAME = {
    1: 'Jan',
    2: 'Feb',
    3: 'Mar',
    4: 'Apr',
    5: 'May',
    6: 'Jun',
    7: 'Jul',
    8: 'Aug',
    9: 'Sep',
    10: 'Oct',
    11: 'Nov',
    12: 'Dec',
} as const satisfies Record<Month, string>;

export function formatDate(
    value: Date | Iso8601Date | UnixTimestampInSeconds,
    dateFormat: DateFormat,
) {
    const dateTime = toDate(value);

    const year = dateTime.getFullYear();
    const month = (dateTime.getMonth() + 1) as Month;
    const day = dateTime.getDate();

    const monthName = MONTH_NAME[month];

    return replace(dateFormat, {
        [`YYYY`]: year,
        [`YY`]: year % 100,
        [`MMM`]: monthName, // FIXME: Add i18n for these
        [`MM`]: month < 10 ? `0${month}` : `${month}`,
        [`M`]: month,
        [`DD`]: day < 10 ? `0${day}` : `${day}`,
        [`D`]: day,
    });
}

export function formatTime(
    value: Date | Iso8601Date | UnixTimestampInSeconds,
    timeFormat: TimeFormat,
) {
    const dateTime = toDate(value);

    const hours = dateTime.getHours();

    return replace(timeFormat, {
        [`HH`]: hours,
        [`hh`]: hours === 0 ? 12 : hours % 12,
        [`mm`]: dateTime.getMinutes(),
        [`a`]: hours === 0 || hours < 12 ? 'am' : 'pm', // FIXME: Add i18n for these maybe?
    });
}

export function FormattedDate({
    format = DEFAULT_DATE_FORMAT,
    value,
    ...attributes
}: FormattedDate.Props) {
    const dateTime = toDate(value);

    return (
        <time {...attributes} dateTime={dateTime.toISOString()}>
            {formatDate(dateTime, format)}
        </time>
    );
}

export namespace FormattedDate {
    export type Props = {
        value: Date | Iso8601Date | UnixTimestampInSeconds;
        format?: DateFormat;
    } & Omit<TimeHTMLAttributes<HTMLTimeElement>, 'dateTime'>;
}

export function FormattedTime({
    format = DEFAULT_TIME_FORMAT,
    value,
    ...attributes
}: FormattedTime.Props) {
    const dateTime = toDate(value);

    return (
        <time {...attributes} dateTime={dateTime.toISOString()}>
            {formatTime(dateTime, format)}
        </time>
    );
}

export namespace FormattedTime {
    export type Props = {
        value: Date | Iso8601Date | UnixTimestampInSeconds;
        format?: TimeFormat;
    } & Omit<TimeHTMLAttributes<HTMLTimeElement>, 'dateTime'>;
}

function replace(text: string, replacements: Record<string, string | number>): string {
    const pattern = new RegExp(
        Object.keys(replacements)
            .sort((a, b) => -cmp(a.length, b.length))
            .join('|'),
        'g',
    );

    return text.replace(pattern, (substring) => String(replacements[substring]));
}

export function toDate(value: Date | Iso8601Date | UnixTimestampInSeconds): Date {
    if (typeof value === 'string') {
        return new Date(value);
    }

    if (typeof value === 'number') {
        return new Date(value * 1000);
    }

    return value;
}

function cmp(a: number, b: number) {
    if (a === b) return 0;
    return a < b ? -1 : 1;
}
