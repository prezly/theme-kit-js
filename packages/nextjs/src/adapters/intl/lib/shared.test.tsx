import renderer from 'react-test-renderer';

import { formatMessageFragment, formatMessageString } from './shared';
import type { IntlDictionary } from './types';

const INTL: IntlDictionary = {
    [`hello.world`]: [
        {
            type: 0,
            value: '¡Hola, ',
        },
        {
            type: 1,
            value: 'name',
        },
        {
            type: 0,
            value: '!',
        },
    ],
};

describe('formatMessageString', () => {
    it('should format intl-generated message intl object using the given values', () => {
        const formatted = formatMessageString(
            INTL,
            { id: 'hello.world', defaultMessage: 'Hello, {name}!' },
            { name: 'Julio' },
        );

        expect(formatted).toBe('¡Hola, Julio!');
    });

    it('should leave placeholders if the given values object does not have enough data', () => {
        const formatted = formatMessageString(
            INTL,
            { id: 'hello.world', defaultMessage: 'Hello, {name}!' },

            {},
        );

        expect(formatted).toBe('¡Hola, {name}!');
    });

    it('should fallback to the default message if the dictionary does not have it', () => {
        const formatted = formatMessageString(
            {},
            { id: 'hello.world', defaultMessage: 'Hello, {name}!' },
            { name: 'Julio' },
        );

        expect(formatted).toBe('Hello, Julio!');
    });
});

describe('formatMessageFragment', () => {
    it('should format intl-generated message intl object using the given values', () => {
        const formatted = formatMessageFragment(
            INTL,
            { id: 'hello.world', defaultMessage: 'Hello, {name}!' },
            { name: <strong>Julio</strong> },
        );

        expect(renderer.create(formatted).toJSON()).toMatchSnapshot();
    });

    it('should leave placeholders if the given values object does not have enough data', () => {
        const formatted = formatMessageFragment(
            INTL,
            { id: 'hello.world', defaultMessage: 'Hello, {name}!' },
            {},
        );

        expect(renderer.create(formatted).toJSON()).toMatchSnapshot();
    });

    it('should fallback to the default message if the dictionary does not have it', () => {
        const formatted = formatMessageFragment(
            {},
            { id: 'hello.world', defaultMessage: 'Hello, {name}!' },
            { name: <strong>Julio</strong> },
        );

        expect(renderer.create(formatted).toJSON()).toMatchSnapshot();
    });
});
