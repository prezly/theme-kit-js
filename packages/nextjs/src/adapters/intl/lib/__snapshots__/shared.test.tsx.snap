// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatMessageFragment should fallback to the default message if the dictionary does not have it 1`] = `
[
  "Hello, ",
  <strong>
    Julio
  </strong>,
  "!",
]
`;

exports[`formatMessageFragment should format intl-generated message intl object using the given values 1`] = `
[
  "¡Hola, ",
  <strong>
    Julio
  </strong>,
  "!",
]
`;

exports[`formatMessageFragment should leave placeholders if the given values object does not have enough data 1`] = `
[
  "¡Hola, ",
  "{name}",
  "!",
]
`;
